File: Lab 5/MATLAB/Task_4.m
clear

[message,fs] = audioread('Music.wav');
time_endpt = floor(length(message)/fs);
message_noisy = message;

for T = 0:time_endpt - 1

    % Defining all the common parameters for each second
    start_time = 0;
    stop_time = 1;
    ts = 1/fs;
    time = start_time:ts:stop_time;

    % Generating the message signal
    message_t_left = message((fs*(T)+1):(fs*(T+1)+1),1);
    message_t_right = message((fs*(T)+1):(fs*(T+1)+1),2);

    figure(1)
    subplot(1,2,1)
    plot(time + T, message_t_left)
    title('Message - Time Domain')
    xlabel('Time')
    ylabel('Amplitude')
    hold on
    
    subplot(1,2,2)
    plot(time + T, message_t_right)
    title('Message, Right - Time Domain')
    xlabel('Time')
    ylabel('Amplitude')
    hold on
    
    % Modelling the Channel
    % The bandwidth should be sufficiently large to encompass all the
    % frequencies within the music sample.
    B = 20000;
    channel_t = 2*B*sinc(2*pi*B*(time - (- start_time + stop_time)/2));
    
    % Modelling noise
    mu = 0;  % Sum of last two digits of ID
    sigma_square = 5;  % Sum of last three digits of ID
    sigma = sqrt(sigma_square);

    noise = mu + sigma * randn(numel(message_t_left),1);
    
    % There can be two implementations to this which work - when using the
    % following segment, uncomment the delta function implementation.
    output_t_left = conv(message_t_left, channel_t,'same')/fs + 0.01*noise;
    output_t_right = conv(message_t_right, channel_t,'same')/fs + 0.01*noise;
    
    message_noisy((fs*(T)+1):(fs*(T+1)+1), 1) = output_t_left;
    message_noisy((fs*(T)+1):(fs*(T+1)+1), 2) = output_t_right;
    
    figure(2)
    subplot(1,2,1)
    plot(time + T, output_t_left)
    title('Output, Left - Time Domain')
    xlabel('Time')
    ylabel('Amplitude')
    hold on
    
    subplot(1,2,2)
    plot(time + T, output_t_right)
    title('Output, Right - Time Domain')
    xlabel('Time')
    ylabel('Amplitude')
    hold on
end
% sound(message_noisy, fs)
filename = 'Noisy Message.wav';
audiowrite(filename,message_noisy,fs)

File: Lab 5/Python/Task_4.py
import numpy as np
import matplotlib.pyplot as plt
import math
import wave
from scipy.io import wavfile
from playsound import playsound
import os


# Reading the audio file
file = wave.open('Music.wav')
fs_audio = file.getframerate()
n_frames = file.getnframes()
audio_duration = n_frames/float(fs_audio)

# Use int32 as the datatype while extracting binary data from the audio file, else it loses out information and when it is played, it gets reflected.
data = file.readframes(-1)
message_init = np.frombuffer(data, np.int32)
message_data = message_init
message_noisy = np.array(message_data, dtype='int32')

time_endpt = int(audio_duration)
for T in range(time_endpt):

    # Defining all the common parameters for each second
    start_time = 0
    stop_time = 1
    fs = fs_audio
    ts = 1/fs
    time = np.arange(start_time, stop_time, ts)

    # Generating the message signal
    message_t = message_data[T*fs:fs*(T + 1)]
    # Modelling the Channel - Choose the bandwidth such that all the frequencies are captured.
    B = 20000
    channel_t = 2 * B * np.sinc(2 * B * (time - (-start_time + stop_time)/2))

    plt.figure(1)
    plt.plot(time + T, message_t)
    plt.title('Message - Time Domain')
    plt.xlabel('Time')
    plt.ylabel('Amplitude')

    # Modelling noise
    mu = 0  # Sum of last two digits of ID
    sigma_square = 4  # Sum of last three digits of ID
    sigma = math.sqrt(sigma_square)

    noise = mu + sigma * np.random.randn(len(message_t))

    # Modify the amplitude of the noise signal and see the effect
    # The max(message_data) was multiplied to make the noise significant enough and observe it's effect. 
    # We could've normalised the message_data vector too but that does not play well enough in sound, so keep it unchanged and instead scale up the noise.
    output_t = np.convolve(message_t, channel_t, mode='same')/fs + 0.01*max(message_data)*noise
    message_noisy[T * fs:fs*(T + 1)] = output_t

    plt.figure(2)
    plt.plot(time + T, output_t)
    plt.title('Output - Time Domain')
    plt.xlabel('Time')
    plt.ylabel('Amplitude')

    # plt.pause(1)

plt.show()

dir_name = os.getcwd()
dir_name_new = '{}'.format(dir_name).replace('\\', '/')
filename = "Noisy Music.wav"
dir_name_final = dir_name_new + '/' + filename
wavfile.write(filename, fs_audio, message_noisy)
playsound(dir_name_final)
